src/server/game/Movement/MotionMaster.cpp

```diff
@@ -56,144 +59,465 @@ inline bool IsStatic(MovementGenerator* movement)
+                }
+                else
+                    list.emplace_back(type, ObjectGuid::Empty, std::string());
+                break;
+            default:
+                list.emplace_back(type, ObjectGuid::Empty, std::string());
+                break;
+        }
     }
+
+    return list;
 }
 
-MovementGenerator* MotionMaster::top() const
+MovementSlot MotionMaster::GetCurrentSlot() const
 {
-    ASSERT(!empty());
+    if (!_generators.empty())
+        return MOTION_SLOT_ACTIVE;
+
+    if (_defaultGenerator)
+        return MOTION_SLOT_DEFAULT;
 
-    return _slot[_top];
+    return MAX_MOTION_SLOT;
 }
 
-void MotionMaster::Initialize()
+MovementGenerator* MotionMaster::GetCurrentMovementGenerator() const
 {
-    // clear ALL movement generators (including default)
-    while (!empty())
+    if (!_generators.empty())
+        return *_generators.begin();
+
+    if (_defaultGenerator)
+        return _defaultGenerator.get();
+
+    return nullptr;
+}
+
+MovementGeneratorType MotionMaster::GetCurrentMovementGeneratorType() const
+{
+    if (Empty())
+        return MAX_MOTION_TYPE;
+
+    MovementGenerator const* movement = GetCurrentMovementGenerator();
+    if (!movement)
+        return MAX_MOTION_TYPE;
+
+    return movement->GetMovementGeneratorType();
+}
+
+MovementGeneratorType MotionMaster::GetCurrentMovementGeneratorType(MovementSlot slot) const
+{
+    if (Empty() || IsInvalidMovementSlot(slot))
+        return MAX_MOTION_TYPE;
+
+    if (slot == MOTION_SLOT_ACTIVE && !_generators.empty())
+        return (*_generators.begin())->GetMovementGeneratorType();
+
+    if (slot == MOTION_SLOT_DEFAULT && _defaultGenerator)
+        return _defaultGenerator->GetMovementGeneratorType();
+
+    return MAX_MOTION_TYPE;
+}
+
+MovementGenerator* MotionMaster::GetCurrentMovementGenerator(MovementSlot slot) const
+{
+    if (Empty() || IsInvalidMovementSlot(slot))
+        return nullptr;
+
+    if (slot == MOTION_SLOT_ACTIVE && !_generators.empty())
+        return *_generators.begin();
+
+    if (slot == MOTION_SLOT_DEFAULT && _defaultGenerator)
+        return _defaultGenerator.get();
+
+    return nullptr;
+}
+
+MovementGenerator* MotionMaster::GetMovementGenerator(std::function<bool(MovementGenerator const*)> const& filter, MovementSlot slot) const
+{
+    if (Empty() || IsInvalidMovementSlot(slot))
+        return nullptr;
+
+    MovementGenerator* movement = nullptr;
+    switch (slot)
     {
-        MovementGenerator* curr = top();
-        pop();
-        if (curr)
-            DirectDelete(curr);
+        case MOTION_SLOT_DEFAULT:
+            if (_defaultGenerator && filter(_defaultGenerator.get()))
+                movement = _defaultGenerator.get();
+            break;
+        case MOTION_SLOT_ACTIVE:
+            if (!_generators.empty())
+            {
+                MotionMasterContainer::const_iterator itr = std::find_if(_generators.begin(), _generators.end(), [filter](MovementGenerator const* a) -> bool
```

Comment:
auto (its an iterator)
also you can rewrite it as simply `auto itr = std::find_if(_generators.begin(), _generators.end(), std::ref(filter));` // ref is here to avoid copying the function into find_if

Reply:
the function is referenced, there was no copy there, right?
eitherway, much cleaner your way 👍

Reply:
There is a copy, default lambda capture mode is by value, it does not matter what the actual captured type is (you would need capture by ref, [&filter])

Reply:
reference the reference, the more you know 💃